### PR TITLE
LCH-6226: PQ command to identify non-translatable entities.

### PR DESCRIPTION
**Pull request by narendradesai**
Issue: https://backlog.acquia.com/browse/LCH-6226

### CODE REVIEW

**Review by saxumVermes — CHANGES_REQUESTED** [review]
Specifying the entity type is a nice touch, but we want the command to explore all non translatable entities that we don't handle specifically. You can actually ask it from the HandlerRegistry which ones that we do handle and you can mark in a different colour.

A tip can be added if they want to change the behaviour we can link the documentation.

File: src/Command/PqCommands/ContentHubPqNonTranslatables.php [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        $result->setIndicator(
+          $kriName,
+          '',
+          'No non-translatable entities detected.'
+        );
+        return 0;
+      }
+      $entities = $this->loadEntities($etm, $entityType);
+      $nonTranslatables = [];
+      foreach ($entities as $entity) {
```
Comment (saxumVermes) [inline src/Command/PqCommands/ContentHubPqNonTranslatables.php:82]: Wondering if we really need to go through all entities to get whether an entity is translatable or not. I'd be interested in the fact whether a bundle of a given entity type is translatable or not.
If I have a content entity type and has bundle, which has a language_content_setting assigned to it, and content_translation is enabled there, we know that the bundle is translatable therefore the entity itself as well.
If it doesn't have a language_content_setting, or it is disabled it means the entity type is non-translatable.

File: src/Command/PqCommands/ContentHubPqNonTranslatables.php [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    $this->addOption('entity-type', 'e', InputOption::VALUE_OPTIONAL, 'Run checks for the provided entity type. Example node,user,paragraph', 'node');
+    $this->addUsage('ach:pq:non-translatables --entity-type "node,paragraph"');
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * @throws \Exception
+   */
+  protected function runCommand(InputInterface $input, PqCommandResult $result): int {
```
Comment (saxumVermes) [inline src/Command/PqCommands/ContentHubPqNonTranslatables.php:60]: We'd only want to go through content entities.

**Comment by amangrover90** [comment]
> 

@saxumVermes but PQ commands were written with the assumption that the CH module might not be installed, isn't it? It should only need depcalc, shouldn't it?

File: src/Command/PqCommands/ContentHubPqNonTranslatables.php (hunk 1 above)
Comment (narendradesai) [reply src/Command/PqCommands/ContentHubPqNonTranslatables.php]: fixed in [6f47d59](https://github.com/acquia/contenthub-console/pull/35/commits/6f47d59a503068b9c275aabd01d10f35820bec72)

File: src/Command/PqCommands/ContentHubPqNonTranslatables.php (hunk 2 above)
Comment (narendradesai) [reply src/Command/PqCommands/ContentHubPqNonTranslatables.php:60]: fixed in [6f47d59](https://github.com/acquia/contenthub-console/pull/35/commits/6f47d59a503068b9c275aabd01d10f35820bec72)

**Comment by saxumVermes** [comment]
> @saxumVermes but PQ commands were written with the assumption that the CH module might not be installed, isn't it? It should only need depcalc, shouldn't it?

Yeah, that's actually true! If the handler exist we can ask from it, but otherwise it would be just a list. The documentation provides details about what entities we support.

**Comment by narendradesai** [comment]
@saxumVermes added check for HandlerRegistry in https://github.com/acquia/contenthub-console/pull/35/commits/4150db0daa004b82c0c9a8d6a05b62722ef8f137

File: src/Command/PqCommands/ContentHubPqNonTranslatables.php [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    /** @var \Drupal\Core\Entity\EntityTypeManagerInterface $etm */
+    $etm = $this->serviceFactory->getDrupalService('entity_type.manager');
+    /** @var \Drupal\Core\Entity\EntityTypeBundleInfoInterface $bundle_info */
+    $bundle_info = $this->serviceFactory->getDrupalService('entity_type.bundle.info');
+    $entity_type_definitions = $etm->getDefinitions();
+    $entityTypes = empty($entityTypeOption) ? array_keys($entity_type_definitions) : explode(',', $entityTypeOption);
+    if ($this->serviceFactory->hasDrupalService('acquia_contenthub_translations.nt_entity_handler.registry')) {
+      /** @var \Drupal\acquia_contenthub_translations\EntityHandler\HandlerRegistry $entity_handler_registery */
+      $entity_handler_registery = $this->serviceFactory->getDrupalService('acquia_contenthub_translations.nt_entity_handler.registry');
+      $this->handledEntities = array_merge($this->handledEntities, array_keys($entity_handler_registery->getHandlerMapping()));
```
[the remark this answers is not in the document]
Comment (narendradesai) [reply src/Command/PqCommands/ContentHubPqNonTranslatables.php:78]: @saxumVermes are you sure? This is what I get when I do: 
`drush @ds cget acquia_contenthub_translations.settings   `

```
_core:
  default_config_hash: 0c3wduklXpuFzW-tfvtVqBgnjRDsqXhwxVec9uCT6i4
selective_language_import: true
override_translation: false
nt_entity_registry:
  unspecified: {  }
  handler_mapping:
    paragraph: removable
```

File: src/Command/PqCommands/ContentHubPqNonTranslatables.php [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+   * {@inheritDoc}
+   *
+   * @throws \Exception
+   */
+  protected function runCommand(InputInterface $input, PqCommandResult $result): int {
+    $entityTypeOption = $input->getOption('entity-type');
+    /** @var \Drupal\Core\Entity\EntityTypeManagerInterface $etm */
+    $etm = $this->serviceFactory->getDrupalService('entity_type.manager');
+    /** @var \Drupal\Core\Entity\EntityTypeBundleInfoInterface $bundle_info */
+    $bundle_info = $this->serviceFactory->getDrupalService('entity_type.bundle.info');
```
[the remark this answers is not in the document]
Comment (narendradesai) [reply src/Command/PqCommands/ContentHubPqNonTranslatables.php]: Done in [f80d361](https://github.com/acquia/contenthub-console/pull/35/commits/f80d3617ed44a244eee372334b9d75e70e158f5f)

**Review by saxumVermes — APPROVED** [review]

**Pull request merged**